# Step 1 Load & Preview the Dataset

In [1]:
import pandas as pd 

In [2]:
df=pd.read_excel("HW3- dataset.xlsx")
df

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,57,0,0,140,241,0,1,123,1,0.2,1,0,3,0
299,45,1,3,110,264,0,1,132,0,1.2,1,0,3,0
300,68,1,0,144,193,1,1,141,0,3.4,1,2,3,0
301,57,1,0,130,131,0,1,115,1,1.2,1,1,3,0


In [4]:
df.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,63,1,3,145,233,1,0,150,0,2.3,0,0,1,1
1,37,1,2,130,250,0,1,187,0,3.5,0,0,2,1
2,41,0,1,130,204,0,0,172,0,1.4,2,0,2,1
3,56,1,1,120,236,0,1,178,0,0.8,2,0,2,1
4,57,0,0,120,354,0,1,163,1,0.6,2,0,2,1


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0    age      303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


The target variables
0 means there is no presence of heart disease, and 
1  means there is presence of heart disease.

# Step 2 Data Cleanint & Feature Preparation

In [29]:
df.isna().sum()

 age        0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [31]:
# 4 categorical variables
# One-hot encoding
df_encoding=pd.get_dummies(df,columns=['cp','slope', 'ca', 'thal'],dtype=int)

In [32]:
df_encoding

,age,sex,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,target,...,slope_2,ca_0,ca_1,ca_2,ca_3,ca_4,thal_0,thal_1,thal_2,thal_3
0,63,1,145,233,1,0,150,0,2.3,1,...,0,1,0,0,0,0,0,1,0,0
1,37,1,130,250,0,1,187,0,3.5,1,...,0,1,0,0,0,0,0,0,1,0
2,41,0,130,204,0,0,172,0,1.4,1,...,1,1,0,0,0,0,0,0,1,0
3,56,1,120,236,0,1,178,0,0.8,1,...,1,1,0,0,0,0,0,0,1,0
4,57,0,120,354,0,1,163,1,0.6,1,...,1,1,0,0,0,0,0,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
298,57,0,140,241,0,1,123,1,0.2,0,...,0,1,0,0,0,0,0,0,0,1
299,45,1,110,264,0,1,132,0,1.2,0,...,0,1,0,0,0,0,0,0,0,1
300,68,1,144,193,1,1,141,0,3.4,0,...,0,0,0,1,0,0,0,0,0,1
301,57,1,130,131,0,1,115,1,1.2,0,...,0,0,1,0,0,0,0,0,0,1


# Step 3 Baseline Model 1-k-Nearest Neighbors 

In [33]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report,confusion_matrix
X=df.drop("target",axis=1)
y=df["target"]
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25,random_state=42)
scaler=StandardScaler()
X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.transform(X_test)
X_test_scaled

array([[ 0.26745345,  0.71647284, -0.97937092,  1.16826465,  0.55847168,
        -0.382707  , -1.04486089, -1.67196572,  1.46723474, -0.38759507,
        -0.6634347 ,  0.30327986, -2.20282109],
       [ 0.4831264 ,  0.71647284,  1.92041862,  2.34508328,  0.78298125,
        -0.382707  , -1.04486089,  0.41169299, -0.6815542 , -0.75004237,
        -0.6634347 , -0.69446692,  1.14193067],
       [ 0.26745345,  0.71647284,  0.95382211,  1.16826465, -2.24789797,
         2.61296502,  0.83754722,  1.03235729, -0.6815542 , -0.75004237,
         0.97351831,  0.30327986,  1.14193067],
       [ 0.15961698, -1.39572632, -0.97937092,  0.22680976,  3.04678611,
        -0.382707  , -1.04486089,  0.01269452,  1.46723474,  0.79035864,
        -0.6634347 ,  1.30102663,  1.14193067],
       [ 1.77716406, -1.39572632,  0.95382211, -1.18537259,  0.35267124,
         2.61296502, -1.04486089, -0.87396876, -0.6815542 , -0.93126602,
         0.97351831,  0.30327986, -0.53044521],
       [ 0.26745345,  0.716472

In [34]:
knn=KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train_scaled,y_train)
value_prediction=knn.predict(X_test_scaled)
print(classification_report(y_test,value_prediction))

              precision    recall  f1-score   support

           0       0.86      0.89      0.87        35
           1       0.90      0.88      0.89        41

    accuracy                           0.88        76
   macro avg       0.88      0.88      0.88        76
weighted avg       0.88      0.88      0.88        76



In [35]:
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score
accuracy1=accuracy_score(y_test,value_prediction)
precision1=precision_score(y_test,value_prediction)
recall1=recall_score(y_test,value_prediction)
f1_1=f1_score(y_test,value_prediction)
print("Accuracy:",accuracy1)
print("Precision:",precision1)
print("Recall:",recall1)
print("F1-score:",f1_score)

Accuracy: 0.881578947368421
Precision: 0.9
Recall: 0.8780487804878049
F1-score: <function f1_score at 0x149396a20>


In [36]:
confusionmatrix=confusion_matrix(y_test,value_prediction)
confusionmatrix

array([[31,  4],
       [ 5, 36]])

In [37]:
# Threshold 0.2 
y_probabilties=knn.predict_proba(X_test_scaled)[:,1]
threshold=0.2
y_predict_threshold=(y_probabilties >= threshold).astype(int)
print(confusion_matrix(y_test,y_predict_threshold))

[[18 17]
 [ 3 38]]


In this confusion matrix,  31 patients  were correctly predicted as having no heart diease, and 4 were false positve,predicted to have  heart diease when they actually did not. 36 patients  were correctly predicted as having heart diease, while 5 patients predicted as having no heart diease when they actually did. Although the count of 5 is small, it is  critical because these 5 patients do have heart diease. 

# Step 4  Baseline Model 2 -Decision Tree Classifier

In [38]:
from sklearn.tree import DecisionTreeClassifier

In [39]:
tree_pipeline=DecisionTreeClassifier(criterion='gini',max_depth=3,random_state=42)

In [40]:
tree_pipeline.fit(X_train,y_train)

DecisionTreeClassifier(max_depth=3, random_state=42)

In [41]:
y_prediction=tree_pipeline.predict(X_test)

In [42]:
print(classification_report(y_test,y_prediction))


              precision    recall  f1-score   support

           0       0.74      0.80      0.77        35
           1       0.82      0.76      0.78        41

    accuracy                           0.78        76
   macro avg       0.78      0.78      0.78        76
weighted avg       0.78      0.78      0.78        76



In [43]:
accuracy=accuracy_score(y_test,y_prediction)
precision=precision_score(y_test,y_prediction)
recall=recall_score(y_test,y_prediction)
f1=f1_score(y_test,y_prediction)
print("Accuracy:",accuracy)
print("Precision:",precision)
print("Recall:",recall)
print("F1-score:",f1)
 

Accuracy: 0.7763157894736842
Precision: 0.8157894736842105
Recall: 0.7560975609756098
F1-score: 0.7848101265822786


In [44]:
confusionmatrix1=confusion_matrix(y_test,y_prediction)
confusionmatrix1

array([[28,  7],
       [10, 31]])

Decision Trees do not require feature scaling because the algorithm learns by splitting the data into branches based on feature values.The algorithm use these branches to predict whether future patients have heart disease or not. <br>

In contrast, K-NN requires feature scaling because some features have maximum and minmum values that can affect the predictions, such as cholestrol, chest pain, thalach, and slope. After scaling, K-NN predicts by finding the 5 nearest data points and determining which label( 0 for no heart disease, 1 for heart disease) appears most frequently to make the prediction. 

# Step 5 Model Comparison Table

In [45]:
data={
    "Model":["K-NN","Decision Tree"],
    "Accuracy":[0.88,0.77],
    "Precision":[0.9,0.81],
    "Recall":[0.87,0.75],
    "F1":[0.88,0.78]}
df1=pd.DataFrame(data)
print(df1)

           Model  Accuracy  Precision  Recall    F1
0           K-NN      0.88       0.90    0.87  0.88
1  Decision Tree      0.77       0.81    0.75  0.78


The model perfomed better was K-NN because it had an over better classifcation report than Decision Tree and the confusion matrix of K-NN  had an overall less false negative of heart disease patients than Decision Tree confusion matrix.

In [46]:
#Decision tree cross validation to check for overfitting
from sklearn.model_selection import cross_val_score
train_prediction=tree_pipeline.predict(X_train)
train_accuracy=accuracy_score(y_train,train_prediction)
cross_validation_score=cross_val_score(tree_pipeline,X,y,cv=5)
cross_validation_mean=cross_validation_score.mean()
print("Training Accuracy:",train_accuracy)
print("Cross Validation Scores:",cross_validation_score)
print("Average Cross Validation Accuracy:",cross_validation_mean)

Training Accuracy: 0.8634361233480177
Cross Validation Scores: [0.80327869 0.85245902 0.80327869 0.78333333 0.75      ]
Average Cross Validation Accuracy: 0.7984699453551911


In [47]:
#KNN cross validation to check for overfitting
train_prediction_knn=knn.predict(X_train_scaled)
train_accuracy_knn=accuracy_score(y_train,train_prediction_knn)
print("K-NN Training Accuracy:",train_accuracy_knn)
X_scaled=scaler.fit_transform(X)
cross_validation_score_knn=cross_val_score(knn,X_scaled,y,cv=5)
cross_validation_mean_knn=cross_validation_score_knn.mean()
print("Training Accuracy:",train_accuracy_knn)
print("Cross Validation Scores:",cross_validation_score_knn)
print("Average Cross Validation Accuracy:",cross_validation_mean_knn)


K-NN Training Accuracy: 0.8634361233480177
Training Accuracy: 0.8634361233480177
Cross Validation Scores: [0.78688525 0.85245902 0.80327869 0.86666667 0.75      ]
Average Cross Validation Accuracy: 0.8118579234972678


Decision tree shows  slight overfititng, since  the training accuracy is 0.86, while the average cross validation accuracy is 0.79. KNN also shows sligtly overfitting,with a training accuracy of 0.86 and an  average cross validation accuracy of 0.81. Overall, K-NN overfits slightly less than Decision Tree.<br>

The model I would  choose as my official baseline forward will be K-NN,  since its confusion matrix shows fewer false negatives compare to the Decision Tree, and it also shows less overfitting. 

# Step 6 Reflection

For this project,  I learned the fundamentals of using K-Nearest Neighbors, and Decison Trees. K-NN works by selecting a specific number of nearby data points to make a prediction, while a Decison Tree splits the data into branches to make a prediction. K-NN is appropriate when dealing with  several features that contains maximum and minimum values, which need to be scaled before  comparing to the chosen number of data points.Scaling improves the precision for the prediction modeling.

Decison Tree are preferable for classification problems, such as  predicting whether a patient has heart disease in this dataset. It is also  recommended to apply cross validation to evaluate the model's accuracy and identify ways to improve performance metrics.  

The challenges in classification data preparation is ensuring that all categorical features, are correctly encoded and scaled when applying K-NN. For Decision Trees, using a higher max_depth can lead to  overfitting. 